lib/charms/operator_libs_linux/v1/systemd.py

```diff
@@ -66,149 +71,146 @@
+    try:
+        proc = subprocess.run(
+            [*cmd],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            text=True,
+            bufsize=1,
+            encoding="utf-8",
+            check=check,
+        )
+        logger.debug(
+            f"Command {cmd} exit code: {proc.returncode}. systemctl output:\n{proc.stdout}"
+        )
+        return proc.returncode
+    except subprocess.CalledProcessError as e:
+        raise SystemdError(
+            f"Command {cmd} failed with returncode {e.returncode}. systemctl output:\n{e.stdout}"
+        )
 
 
-def _systemctl(
-    sub_cmd: str, service_name: str = None, now: bool = None, quiet: bool = None
-) -> bool:
-    """Control a system service.
+def service_running(service_name: str) -> bool:
+    """Report whether a system service is running.
 
     Args:
-        sub_cmd: the systemctl subcommand to issue
-        service_name: the name of the service to perform the action on
-        now: passes the --now flag to the shell invocation.
-        quiet: passes the --quiet flag to the shell invocation.
+        service_name: the name of the service to check
     """
-    cmd = ["systemctl", sub_cmd]
-
-    if service_name is not None:
-        cmd.append(service_name)
-    if now is not None:
-        cmd.append("--now")
-    if quiet is not None:
-        cmd.append("--quiet")
-    if sub_cmd != "is-active":
-        logger.debug("Attempting to {} '{}' with command {}.".format(cmd, service_name, cmd))
+    # If returncode is 0, this means that is service is active.
+    if _systemctl("--quiet", "is-active", service_name) == 0:
+        return True
     else:
-        logger.debug("Checking if '{}' is active".format(service_name))
+        return False
 
-    proc = subprocess.Popen(cmd, **_popen_kwargs())
-    last_line = ""
-    for line in iter(proc.stdout.readline, ""):
-        last_line = line
-        logger.debug(line)
 
-    proc.wait()
+def service_failed(service_name: str) -> bool:
+    """Report whether a system service has failed.
 
-    if proc.returncode < 1:
+    Args:
+        service_name: the name of the service to check
+    """
+    # If returncode is 0, this means that the service has failed.
+    if _systemctl("--quiet", "is-failed", service_name) == 0:
         return True
-
-    # If we are just checking whether a service is running, return True/False, rather
-    # than raising an error.
-    if sub_cmd == "is-active" and proc.returncode == 3:  # Code returned when service not active.
-        return False
-
-    if sub_cmd == "is-failed":
+    else:
         return False
 
-    raise SystemdError(
-        "Could not {}{}: systemd output: {}".format(
-            sub_cmd, " {}".format(service_name) if service_name else "", last_line
-        )
-    )
-
 
-def service_running(service_name: str) -> bool:
-    """Determine whether a system service is running.
+def service_start(*args: str) -> None:
```

Comment:
Yeah, on reflection I think we probably shouldn't make the change to the return type without a major version bump (and I don't think it's worth doing that). So what about doing something like this for all of these functions.

```python
def service_start(*args: str) -> bool:
    """Start a system service.

    Args:
        args: Name (or names) of services to start.

    Returns:
        On success, this function always returns True (for historical reasons).

    Raises:
        SystemdError: on error calling systemctl.
"""
    _systemctl("start", *args, check=True)
    return True
```

Reply:
Sounds good - I can change it back to returning `True`. Since _systemctl(...) can either return an exit code of 0 or raise a `SystemdError`, we can just have the return line be `return _systemctl("start", *args, check=True) == 0`. I don't like the idea of returning `True` just using `return True` to ensure that we don't break compatibility because it appears that we are just ignoring what `_systemctl(...)` does to the underlying machine.

Reply:
Switched back to returning bools in commit https://github.com/canonical/operator-libs-linux/pull/106/commits/c951f68b8fe18cbab6fc69858e838283063c1e48